Fix support for Intel Compute Runtime with VectorSize > 1

proski opened the pull request:
Fix support for Intel Compute Runtime with VectorSize > 1

The fallback implementation of amd_bitalign() triggers a bug with Intel Compute
Runtime (NEO) versions from 23.22.26516.18 to 24.45.31740.9 inclusive.

https://github.com/intel/intel-graphics-compiler/issues/358

The bug affects all but the first component of the vectors, so the self-tests
would pass with VectorSize=1. For higher values of VectorSize, including the
default VectorSize=2, approximately half of the self-tests fail, all in
barrett32 kernels.

Add generic_bitalign() that is always implemented using shifts. Use it in all
cases when the destination is the same as one of the sources.

If Intel Compute Runtime is detected, use 64-bit shifts in generic_bitalign().
For other platforms, keep using 32-bit shifts.

Make amd_bitalign() an alias to generic_bitalign() on systems where
amd_bitalign() is not available. That way, it would also expand to 64-bit
shifts for Intel Compute Runtime.

proski commented:
Current status:
- The quick test passes both with and without fp64
- The self-test (`-st`) passes without fp64, fails with fp64
- The extended self-test (`-st2`) has not completed yet

Please don't use this PR in production unless it passes the self-test on your system.

proski commented:
Update:
* Preserve the existing choices of whether to use `amd_bitalign()` on platforms where it's available.
* Replace the existing shifts with `amd_bitalign_emulated()` which is always implemented using shifts - even in comments.
* Detect Intel NEO without relying on `GPUType` (it's for optimization, not for workarounds).

TODO:

* More testing
* Probably a better way to detect if the workaround is needed
* Benchmark 32-bit and 64-bit shifts on Intel for VectorSize=1, use whatever is faster

proski commented:
Update:
* rename `amd_bitalign_emulated` to `generic_bitalign`
* update comments accordingly

Confirmed that 64-bit shifts are faster than 32-bit shifts on Intel with `VectorSize=1` - no need to make a special case for `VectorSize=1`.

preda commented on src/common.cl:
```diff
@@ -170,6 +170,22 @@ uint popcount(uint x)
+// generic_bitalign emulates amd_bitalign using shifts. generic_bitalign can be
+// used instead of amd_bitalign if benchmarks show that it's faster.
+#ifdef cl_intel_subgroups
+// Workaround for Intel NEO that miscompiles shifts on uint vectors - use ulong instead
```
Would it be worth to let Intel know about this problem -- maybe they would like to fix it? (independently of our workaround here)

proski replied:
Reported the issue to Intel: https://github.com/intel/compute-runtime/issues/790
I'm glad I could put together a simple demo.
The breakage must have happened between versions 22.14.22890 and 23.43.27642. The precompiled binaries are for Ubuntu and I only have Fedora now, compiling intel-opencl is very time consuming, but I might give it another try later.

proski replied:
I was able to use Ubuntu in WSL. If turns out that 23.17.26241.22 is the last good release and
23.22.26516.18 is the first bad release. Also, the latest release, 24.48.31907.7, fixes the issue. But it's too new and most users don't have it yet.

proski commented:
Not WIP anymore - ready for review.

* Determined the affected versions of Intel Compute Runtime
* Reported the issue to Intel
* Updated the comments accordingly
* Tested on Radeon Pro 560X and found no performance degradation

proski commented:
* Updated the description and a comment in the code - no code changes.
* I was asked to open an issue for the Intel Graphics Compiler, so I changed links to that issue: https://github.com/intel/intel-graphics-compiler/issues/358
* Tested the code on Windows 11 with the current Intel driver - it's also affected and this PR is fixing everything.
